torchtune/utils/ppo_utils/_generation.py

```diff
@@ -0,0 +1,182 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Optional
+
+import torch
+from torchtune.utils._generation import sample, update_stop_tokens_tracker
+
+
+def generate_next_token(
+    model: torch.nn.Module,  # TODO (SalmanMohammadi) leaving as nn.module until TransformerDecoder refactor
+    input_pos: torch.Tensor,
+    x: torch.Tensor,
+    *,
+    mask: Optional[torch.Tensor] = None,
+    temperature: float = 1.0,
+    top_k: int = None,
+) -> torch.Tensor:
+    """Generates the next tokens."""
+    # model produces logits in [bsz, seq_length, vocab_size]
+    # we want to take the last token's logits as the input to the next model call
+    logits = model(x, input_pos=input_pos, mask=mask)[:, -1]
+    return sample(logits, temperature, top_k)
+
+
+def generate_next_token_with_value_head_model(
```

Comment:
Similar comment here -- is the purpose of this just to unpack the tuple returned by the value head model? If so I wonder if it's more appropriate to unpack directly in the recipe with an appropriate code commment

Reply:
Generation utils have now been updated to return logits too. This saves an additional forward pass to calculate logprobs over generated sequences.